tests/integration/conftest.py

```diff
@@ -23,16 +25,23 @@ class TestConfig:
     data_source_config: DataSourceTestConfig
     data: pd.DataFrame
     extra_data: Mapping[str, pd.DataFrame]
+    secondary_source_config: Union[DataSourceTestConfig, None] = None
+    secondary_data: Union[pd.DataFrame, None] = None
 
     @override
     def __hash__(self) -> int:
+        if self.secondary_data is None and self.secondary_source_config is not None:
```

Comment:
Validation like this seems wrong in a hash function. Maybe needed for type narrowing though? I'd also consider (but I'm not strongly pushing for it) another class that holds non-nullable secondary source configs and data, and then making that class optional. Let the type system enforce this for us.